javascript/i18n/phonenumbers/phonenumberutil.js

```diff
@@ -743,47 +743,122 @@ i18n.phonenumbers.PhoneNumberUtil.VALID_PHONE_NUMBER_ =
- * the digits that you want to capture as part of the extension, or else parsing
- * will fail! We allow two options for representing the accented o - the
- * character itself, and one in the unicode decomposed form with the combining
- * acute accent.
+ * Helper initialiser method to create the regular-expression pattern to match
+ * extensions.
  *
- * @const
- * @type {string}
+ * @return {string} constructued RegEx pattern to capture extensions.
  * @private
  */
-i18n.phonenumbers.PhoneNumberUtil.EXTN_PATTERNS_FOR_PARSING_ =
-    i18n.phonenumbers.PhoneNumberUtil.RFC3966_EXTN_PREFIX_ +
-    i18n.phonenumbers.PhoneNumberUtil.CAPTURING_EXTN_DIGITS_ + '|' +
-    '[ \u00A0\\t,]*' +
-    '(?:e?xt(?:ensi(?:o\u0301?|\u00F3))?n?|\uFF45?\uFF58\uFF54\uFF4E?|' +
-    '\u0434\u043E\u0431|' +
-    '[;,x\uFF58#\uFF03~\uFF5E]|int|anexo|\uFF49\uFF4E\uFF54)' +
-    '[:\\.\uFF0E]?[ \u00A0\\t,-]*' +
-    i18n.phonenumbers.PhoneNumberUtil.CAPTURING_EXTN_DIGITS_ + '#?|' +
-    '[- ]+([' + i18n.phonenumbers.PhoneNumberUtil.VALID_DIGITS_ + ']{1,5})#';
+i18n.phonenumbers.PhoneNumberUtil.createExtnPattern_ =
+    function() {
+ // We cap the maximum length of an extension based on the ambiguity of the way
+ // the extension is prefixed. As per ITU, the officially allowed length for
+ // extensions is actually 40, but we don't support this since we haven't seen real
+ // examples and this introduces many false interpretations as the extension labels
+ // are not standardized.
+ /** @type {string} */
+ var extLimitAfterExplicitLabel = '20';
+ /** @type {string} */
+ var extLimitAfterLikelyLabel = '15';
+ /** @type {string} */
+ var extLimitAfterAmbiguousChar = '9';
+ /** @type {string} */
+ var extLimitWhenNotSure = '6';
+
+ /** @type {string} */
+ var possibleSeparatorsBetweenNumberAndExtLabel = "[ \u00A0\\t,]*";
+ // Optional full stop (.) or colon, followed by zero or more spaces/tabs/commas.
+ /** @type {string} */
+ var possibleCharsAfterExtLabel = "[:\\.\uFF0E]?[ \u00A0\\t,-]*";
+ // Here the extension is called out in more explicit way, i.e mentioning it obvious
+ // patterns like "ext.".
+ /** @type {string} */
+ var explicitExtLabels =
+     "(?:e?xt(?:ensi(?:o\u0301?|\u00F3))?n?|\uFF45?\uFF58\uFF54\uFF4E?|\u0434\u043E\u0431|anexo)";
+ // One-character symbols that can be used to indicate an extension, and less
+ // commonly used or more ambiguous extension labels.
+ /** @type {string} */
+ var ambiguousExtLabels = "(?:[x\uFF58#\uFF03~\uFF5E]|int|\uFF49\uFF4E\uFF54)";
+ // When extension is not separated clearly.
+ /** @type {string} */
+ var ambiguousSeparator = "[- ]+";
+ // ",," is commonly used for auto dialling the extension when connected. First
+ // comma is matched through possibleSeparatorsBetweenNumberAndExtLabel, so we do
+ // not repeat it here. Semi-colon works in Iphone and Android also to pop up a
+ // button with the extension number following.
+ /** @type {string} */
+ var autoDiallingAndExtLabelsFound = "(?:,{2}|;)";
```

Comment:
Move to below possibleSeparatorsNumberExtLabelNoComma for consistency with Java

Reply:
Actually in general can you reorder these definitions to match Java?

Reply:
Done.